overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +52,245 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
+
+	return time.Unix(seconds, nanoseconds), nil
+}
+
+func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+	if m.ensuredSnapSvcs {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// we are seeded, now we need to find all snap services and re-generate
+	// services as necessary
+
+	// ensure all snap services are updated
+	allStates, err := snapstate.All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	snapsMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		// TODO: handle vitality rank here too from configcore
+		snapsMap[info] = nil
+	}
+
+	// setup ensure options
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := devicestate.DeviceCtx(m.state, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: should we use an actual interacter here ?
+	modified, err := wrappers.EnsureSnapServices(snapsMap, ensureOpts, progress.Null)
+	if err != nil {
+		return err
+	}
+
+	// if nothing was modified or we are not on UC18+, we are done
+	if len(modified) == 0 || deviceCtx.Classic() || deviceCtx.Model().Base() == "" {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	// otherwise we need to check for all the services that were modified if
+	// they were recently killed when we modified usr-lib-snapd.mount as part of
+	// a snapd snap refresh
+
+	// as a last resort, if we fail in trying to start any services, we should
+	// reboot
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		// TODO: reboot the system immediately here
+	}()
+
+	// we decide on which services to restart by identifying (out of the set of
+	// services we just modified) services that were stopped after
+	// usr-lib-snapd.mount was written, but before usr-lib-snapd.mount was last
+	// stopped - this is the time window in which snapd (accidentally) killed
+	// all snap services using Requires=, see LP #1924805 for full details, so
+	// we need to undo that by restarting those snaps
+
+	// TODO: use the var from core18.go here instead
+	st, err := os.Stat("/etc/systemd/system/usr-lib-snapd.mount")
+	if err != nil {
+		return err
+	}
+
+	// TODO: we should check if usr-lib-snapd.mount was modified before the
+	// current boot time, if it was then we can just skip this since we know
+	// any service stops that happened were unrelated
+	lowerTimeBound := st.ModTime()
+
+	// connect to the system bus to talk to systemd
+	conn, err := dbus.SystemBus()
+	if err != nil {
+		panic(err)
```

Comment:
why the panic here?

Reply:
Sorry leftover from my helper executable I was testing with which wasn't handling errors, this should be returning errors normally